File: solutions/fsharp/perfect-numbers/1/PerfectNumbers.fs
module PerfectNumbers

type Classification =
    | Perfect
    | Abundant
    | Deficient

let classify =
    function
    | n when n < 1 -> None
    | n ->
        [ 1 .. n / 2 ]
        |> Seq.filter (fun i -> n % i = 0)
        |> Seq.sum
        |> (function
        | s when s < n -> Deficient
        | s when s > n -> Abundant
        | s when s = n -> Perfect
        | s -> failwith $"Miraculously invalid number {s}")
        |> Some

File: solutions/csharp/perfect-numbers/1/PerfectNumbers.cs
using System;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    private static int AliquotSum (int number) => Enumerable.Range(1, (number - 1))
          .Aggregate(0, (total, n) => number % n == 0 ? n + total : total);
    public static Classification Classify(int number)
    {
        var aliquotSum = AliquotSum(number);

        return aliquotSum > number
          ? Classification.Abundant
          : aliquotSum == number
          ? Classification.Perfect
          : Classification.Deficient;
    }
}
